### PR TITLE
[Managed Iceberg] add GiB autosharding

### PR DESCRIPTION
**Pull request by ahmedabu98**
Adds auto-sharding to Iceberg streaming writes using GroupIntoBatches.

In streaming writes, bundles are often very small and can even be single-elements. We write each bundle to a file, so this can behavior can lead to many small files.

To solve this, we group records into batches set by a triggering frequency (as well as record and byte size limits). Now, the number of written data files is more easily controlled. Essentially, every triggering frequency duration, roughly N data files are written, where N is the number of workers. To decrease the number of written files, one can increase their triggering frequency or reduce their parallelism.

### CODE REVIEW

**Comment by ahmedabu98** [comment]
assign set of reviewers

**Review by chamikaramj — COMMENTED** [review]
Thanks. LGTM.

**Comment by chamikaramj** [comment]
cc: @Naireen @dustin12 in case Dataflow streaming team has additional comments on this.

File: sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/WriteToDestinations.java [hunk 1]
```diff
@@ -19,39 +19,34 @@
...
 import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.base.Preconditions;
 import org.checkerframework.checker.nullness.qual.Nullable;
 import org.joda.time.Duration;
 
-class WriteToDestinations extends PTransform<PCollection<Row>, IcebergWriteResult> {
+class WriteToDestinations extends PTransform<PCollection<KV<String, Row>>, IcebergWriteResult> {
 
-  static final long DEFAULT_MAX_BYTES_PER_FILE = (1L << 40); // 1TB
+  // Used for auto-sharding in streaming. Limits number of records per batch/file
+  private static final int FILE_TRIGGERING_RECORD_COUNT = 100_000;
```
[the remark this answers is not in the document]
Comment (ahmedabu98) [reply sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/WriteToDestinations.java:46]: It's taken from WriteFiles:
https://github.com/apache/beam/blob/6a7ffa587df5eac1e3a3886152891f750638dc85/sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java#L149-L152

[the remark this answers is not in the document]
Comment (ahmedabu98) [reply sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/WriteToDestinations.java:46]: BigQuery batch loads is similar but has a greater record count limit (500,000): 
https://github.com/apache/beam/blob/6a7ffa587df5eac1e3a3886152891f750638dc85/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BatchLoads.java#L120-L126

**Comment by chamikaramj** [comment]
BTW this should block the release IMO since it's an update incompatible change on top of another unreleased update incompatible change.

File: sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/WriteToDestinations.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Naireen) [reply sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/WriteToDestinations.java:46]: It might be a good idea in a follow up PR to expose record and byte count, in case the user wants more flexibility. Also not sure if we want this current default of 100000 to be different from the old default of TRIGGERING_RECORD_COUNT=50,000

File: sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/WriteToDestinations.java [hunk 2]
```diff
@@ -66,7 +61,58 @@ class WriteToDestinations extends PTransform<PCollection<Row>, IcebergWriteResul
...
+
+    // Commit files to tables
+    PCollection<KV<String, SnapshotInfo>> snapshots =
+        writtenFiles.apply(new AppendFilesToTables(catalogConfig));
+
+    return new IcebergWriteResult(input.getPipeline(), snapshots);
+  }
+
+  private PCollection<FileWriteResult> writeTriggered(PCollection<KV<String, Row>> input) {
+    checkArgumentNotNull(
```
Comment (Naireen) [inline sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/WriteToDestinations.java:79]: Why was this removed from the IO, and moved into the transform? Wouldn't it have been better in the IO layer (we could fail earlier?)

Comment (ahmedabu98) [reply sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/WriteToDestinations.java:79]: I moved it here because I thought it'd be cleaner to have all the streaming-related code in one place

We will fail early either way, this check still happens during pipeline construction time.

**Comment by liferoad** [comment]
Let us update CHANGES.md

File: sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/WriteToDestinations.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (chamikaramj) [reply sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/WriteToDestinations.java:46]: I think for ManagedIO in general, it might be good to limit the number of knobs we expose. The idea is for Beam/runner to find reasonable optimal values and manage it on behalf of users.

[the remark this answers is not in the document]
Comment (ahmedabu98) [reply sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/WriteToDestinations.java:46]: +1 to not exposing it (at least not from the get-go)

> Also not sure if we want this current default of 100000 to be different from the old default of TRIGGERING_RECORD_COUNT=50,000

Before a recent PR (#32451), the old default actually wasn't used anywhere. This IO is still pretty new and we haven't stress tested it yet to see what's most optimal. I figured a good starting point would be to follow WriteFiles (100,000) because it's essentially the same function.

**Comment by robertwb** [comment]
Does this introduce an additional shuffle (and if so, are we OK with that)?

**Comment by ahmedabu98** [comment]
@robertwb does GroupIntoBatches count as a shuffle? if so then yeah that's the cost here

I think it's a better alternative than writing a huge amount of small files though -- the difference is pretty noticeable. We also use GroupIntoBatches for other performant IOs (all BigQueryIO writes, FileIO, TextIO). Specifically for Iceberg, the table format's query planning is sensitive to the number of files. Some references: [[1]](https://blog.cloudera.com/optimization-strategies-for-iceberg-tables/#:~:text=Problem%20with%20small%20files), [[2]](https://medium.com/ancestry-product-and-technology/solving-the-small-file-problem-in-iceberg-tables-6c31a295f724), [[3]](https://medium.com/@huwng_learn_things/some-ways-to-solve-small-file-problems-in-apache-iceberg-0ee3014cd4bb).

I was thinking of instead adding a step (after file writes) that merges files together. Iceberg does provide a Spark operation that merges files across the entire table ([compaction](https://iceberg.apache.org/docs/1.5.1/maintenance/#compact-data-files)), but I couldn't find anything more light-weight.

**Comment by chamikaramj** [comment]
I don't think Beam "GroupIntoBatches" introduces a shuffle but I suspect Dataflow would introduce a shuffle/re-shard to make auto-sharding work (not sure how costly that is). Agree with Ahmed that benefits here seems to outweigh the associated cost. We did something very similar to BQ streaming sink to reduce the number of output streams (went from a manually configured 50 shards to auto-sharding). In practice I think, either we would introduce the sharding here or customers would have to add that manually to their pipelines. I prefer the former.

**Comment by robertwb** [comment]
OK, we can go with that.
